File: hw/nase/u3/hw25TemplateMethod/TemplateMethod.java/CaffeineBeverage/src/main/java/espe/edu/ec/caffeinebeverage/model/CaffeineBeverage.java
/*
 * Click nbfs://nbhost/SystemFileSystem/Templates/Licenses/license-default.txt to change this license
 * Click nbfs://nbhost/SystemFileSystem/Templates/Classes/Class.java to edit this template
 */
package espe.edu.ec.caffeinebeverage.model;

/**
 *
 * @author Jennyfer Nase 
 */
public abstract class CaffeineBeverage {
    
    public final void prepareRecipe() {
        boilWater();
        brew();
        pourInCup();
        if (wantsCondiments()) { 
            addCondiments(); 
        }
    }
    
    void boilWater() { 
        System.out.println("Boiling water"); 
    }
    
    void pourInCup() { 
        System.out.println("Pouring into cup"); 
    }
    
    protected abstract void brew();
    protected abstract void addCondiments();
    
    protected boolean wantsCondiments() { 
        return true; 
    }
}

File: hw/tipan/u3/hw26TemplateMethod/TemplateMethod/TemplateMethodJAVASCRIPT/model/CaffeineBeverage.js
class CaffeineBeverage {

    prepareRecipe() {
        this.boilWater();
        this.brew();
        this.pourInCup();

        if (this.wantsCondiments()) {
            this.addCondiments();
        }
    }

    boilWater() {
        console.log("Boiling water");
    }

    brew() {
        throw new Error("The brew() method must be implemented");
    }

    pourInCup() {
        console.log("Pouring into cup");
    }

    addCondiments() {
        throw new Error("The addCondiments() method must be implemented");
    }

    wantsCondiments() {
        return true;
    }
}

module.exports = CaffeineBeverage;